Corrected query and placeholder data for ZH ACL check function

samlikins opened the pull request:
This bug causes the modules to fail displaying.

bradymiller commented on library/acl.inc:
```diff
@@ -787,7 +784,7 @@
                           ON group_settings.section_id = acl_sections.section_id
                       WHERE
-                        acl_sections.`section_identifier` = ? AND group_settings.group_id IN (".implode('?,',$arr_temp).") AND group_settings.allowed = ?";
+                        acl_sections.`section_identifier` = ? AND group_settings.group_id IN (?) AND group_settings.allowed = ?";
```
I am pretty sure this won't work across different sql connectors(adodb, pdo, etc), which is important since may change to PDO at some point:
http://stackoverflow.com/questions/920353/can-i-bind-an-array-to-an-in-condition

bradymiller commented:
Hi Sam,
I am not having this issue on my local instance or on the development demo?(Zend is working fine)
Is this issue mysql version or os specific?
-brady

tmccormi commented:
mi2@mii20:~$ mysql --version
mysql  Ver 14.14 Distrib 5.5.44, for debian-linux-gnu (x86_64) using readline 6.3

bradymiller commented:
What errors were you seeing?

tmccormi commented:
This was in response to Harleys post.

Tony McCormick
www.mi-squared.com
On Dec 23, 2015 3:48 PM, "Brady Miller" <notifications@github.com> wrote:

> What errors were you seeing?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openemr/openemr/pull/88#issuecomment-167010231>.
>

-- 
Please be aware that e-mail communication can be intercepted in 
transmission or misdirected. Please consider communicating any sensitive 
information by telephone.  The information contained in this message may be 
privileged and confidential. If you are NOT the intended recipient, please 
notify the sender immediately with a copy to hipaa-security@mrsb-ltd.com and 
destroy this message.

tmccormi commented:
Input Array does not match ?: SELECT COUNT(allowed) AS count FROM
module_acl_group_settings AS group_settings LEFT JOIN module_acl_sections
AS acl_sections ON group_settings.section_id = acl_sections.section_id
WHERE acl_sections.`section_identifier` = 'ccr' AND group_settings.group_id
IN ('12,11'0,

ERROR: query failed: SELECT COUNT(allowed) AS count FROM
module_acl_group_settings AS group_settings LEFT JOIN module_acl_sections
AS acl_sections ON group_settings.section_id = acl_sections.section_id
WHERE acl_sections.`section_identifier` = ? AND group_settings.group_id IN
(??,?) AND group_settings.allowed = ?

Error: You have an error in your SQL syntax; check the manual that
corresponds to your MySQL server version for the right syntax to use near
'0,' at line 8

/opt/www/vhosts/openemr.mi-squared.com/mu2/openemr/library/acl.inc at
803:sqlQuery
/opt/www/vhosts/
openemr.mi-squared.com/mu2/openemr/interface/main/left_nav.php at
1366:zh_acl_check(1,ccr)

Tony McCormick, CTO
www.mi-squared.com
Support: 866-735-0897, Direct: 713-574-6709
My Calendar:  http://bit.ly/XznvDo
“I am and always will be the optimist. The hoper of far-flung hopes and the
dreamer of improbable dreams.” — The Doctor, Season 6



On Wed, Dec 23, 2015 at 5:40 PM, Tony McCormick <tony@mi-squared.com> wrote:

> This was in response to Harleys post.
>
> Tony McCormick
> www.mi-squared.com
> On Dec 23, 2015 3:48 PM, "Brady Miller" <notifications@github.com> wrote:
>
>> What errors were you seeing?
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/openemr/openemr/pull/88#issuecomment-167010231>.
>>
>

-- 
Please be aware that e-mail communication can be intercepted in 
transmission or misdirected. Please consider communicating any sensitive 
information by telephone.  The information contained in this message may be 
privileged and confidential. If you are NOT the intended recipient, please 
notify the sender immediately with a copy to hipaa-security@mrsb-ltd.com and 
destroy this message.

bradymiller commented:
Hi Sam,

What are your thoughts on using this fix instead(using a placemaker for each variable rather than one for the entire array; also stays away from the implode stuff and just manages it using arrays):
https://github.com/bradymiller/openemr/commit/0718ff564d34da3cc2ed23b73c26f152c996a86a

-brady

samlikins commented:
​
Brady,

I think that was the original intention, however it wasn't being
implemented near properly to do that, nor is it necessary for even PDO.
The implode was necessary to get the placeholders in the query originally,
my implementation removed an unnecessary for loop and removed the implode
for the placeholders. SQL expects a comma separated list of values for
"IN", which can be encapsulated in quotes; as so a single placeholder is
sufficient and more efficient. The previous attempt used two loops and two
implodes to generate that portion of the query, which adds a few
unnecessary opp codes with no added benefits; additionally an implode is
much more efficient than a loop by far. This code is already used in loops,
making it more complex than needed increases its CRAP score needlessly. If
I'd seen any merit in using multiple placeholders I would have fixed the
loop instead of removing it.


Thanks,


Sam Likins

WSI-Services
Owner / Creative Developer
Making Things Work
*727.490.WSIS (9747)*

On Sat, Dec 26, 2015 at 6:56 AM, Brady Miller <notifications@github.com>
wrote:

> Hi Sam,
>
> What are your thoughts on using this fix instead(using a placemaker for
> each variable rather than one for the entire array; also stays away from
> the implode stuff and just manages it using arrays):
> bradymiller@0718ff5
> <https://github.com/bradymiller/openemr/commit/0718ff564d34da3cc2ed23b73c26f152c996a86a>
>
> -brady
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openemr/openemr/pull/88#issuecomment-167318759>.
>

bradymiller commented:
Hi Sam,

What happens if a value within the IN contains a comma?

I also think this will break the logging engine and display (when dealing with binded values) in Administration->Logs?

I reorganized things a bit to decrease the CRAP factor here(I am traveling today, so have not been able to test it) :)
https://github.com/bradymiller/openemr/commit/b1abff5e9c5d2bd374dc08b2150de35133b3edbb

-brady

bradymiller commented:
Hi,
My revised commit above is testing well, and since no reply, have committed it(need to get the bug fix out before the 4.3.0 release).
-brady
